MTM-59757, modified documentation of context support, user scope, and… (#3001)
SoftwareAG/c8y-docs: content/microservice-sdk/java-bundle/developing-microservice.md

@Autowired
private EventApi eventApi;

public List<EventRepresentation> getEvents () {
  final ArrayList<EventRepresentation> events = new ArrayList<>();
  contextService.runWithinContext(contextService.getContext(), () -> {
    eventApi.getEvents().get().getEvents().forEach(event -> events.add(event));
  });
  return events;
}
```

> To demonstrate usage of `ContextService` with simple code examples above, `eventApi` and `userEventApi` have been used while they are already attached to the desired context.

Reviewer: ```suggestion
To demonstrate the usage of `ContextService` with simple code examples above, `eventApi` and `userEventApi` have been used while they are already attached to the desired context.
```

> In more complex applications, care must be taken to not use Platform API related beans with tenant scope in a user scope context and vice versa.

Reviewer 2: Not sure whether this is clear to the customer.
But I'm not sure whether it is reasonable to show an example how not to do it...!?

Author: The previous version of the documentation showed usage of the tenantEventApi in the userScope which was wrong and created unexpected results - apart from the fact that the examples did not work at all.
It seems to be a source of trouble to mix up the scopes.
I would appreciate any (short, few lines) example demonstrating contexts with beans which already are in a certain context.

Author: Please verify in replacing PR https://github.com/Cumulocity-IoT/c8y-docs/pull/3107 , thanks

Within the user scope, the created beans use the credentials of the authenticated user sending the request instead of the default service user for the communication with the platform.

> There are several predefined Platform API related beans with this naming pattern

Reviewer: ```suggestion
There are several predefined Platform API related beans with this naming pattern:
```

> There are several predefined Platform API related beans with this naming pattern:

Reviewer: ```suggestion
There are several predefined Platform API related beans with the following naming pattern:
```

> * inventoryApi - in tenant scope

Reviewer 2: ```suggestion
* <library-name>Api - in tenant scope like inventoryApi
```

> * tenantInventoryApi - in tenant scope

Reviewer 2: ```suggestion
* tenant<library-name>Api - in tenant scope like tenantInventoryApi
```

> * userInventoryApi - in user scope

Reviewer 2: ```suggestion
* user<library-name>Api - in user scope like userInventoryApi
```

These beans are
* inventoryApi
* identityApi
* measurementApi
* deviceControlApi
* alarmApi
* eventApi
* eventBinaryApi
* auditRecordApi
* deviceCredentialsApi
* binariesApi
* userApi
* tenantOptionApi
* systemOptionApi
* tokenApi
* notificationSubscriptionApi

> 

Reviewer 3: Can we instead have a table with three columns listing the full names of each bean instead having users apply patterns to derive the correct names themselves?

Author: A table ahs been added. Please verify in replacing PR https://github.com/Cumulocity-IoT/c8y-docs/pull/3107 , thanks

### Microservice security {#microservice-security}

> The `@EnableMicroserviceSecurity` annotation sets up the standard security configuration for microservices. It requires basic authorization for all endpoints (except for health check endpoint configured using `@EnableHealthIndicator`). A developer can secure its endpoints using standard Spring security annotations, for example, `@PreAuthorize("hasRole('ROLE_A')")` and user's permissions will be validated against user's roles stored on the platform.

Reviewer 3: Is "requires basic authentication" correct here? Can't you use OAI-Secure to invoke a microservice as well?

Author: Microservice security is desribed in more detail in https://cumulocity.com/docs/microservice-sdk/general-aspects/#authentication-and-authorization 
"... A request to a microservice can be authenticated using various authentication mechanisms supported by Cumulocity such as:
- Basic authentication
- OAI-Secure
- SSO
... "
A link to this chapter will be added here.

Author: Please verify in replacing PR https://github.com/Cumulocity-IoT/c8y-docs/pull/3107 , thanks
